lib/web/resources_test.go

```diff
@@ -429,6 +435,311 @@ func TestListResources(t *testing.T) {
 	}
 }
 
+func TestListResourcesPing(t *testing.T) {
+	m := &mockedResourceAPIGetter{}
+
+	m.mockListResources = func(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error) {
+		return nil, trace.NotImplemented("not implemented")
+	}
+
+	// Test v8 ping.
+	mockPingCalled := 0
+	m.mockPing = func(ctx context.Context) (proto.PingResponse, error) {
+		mockPingCalled += 1
+		return proto.PingResponse{ServerVersion: "8.3.1"}, nil
+	}
+	mockHTTPReq, err := http.NewRequest("", "", nil)
+	require.NoError(t, err)
+	_, err = listResources(m, mockHTTPReq, types.KindNode)
+	require.True(t, trace.IsNotImplemented(err))
+	require.Equal(t, 1, mockPingCalled)
+
+	// Test v9.0 ping.
+	mockPingCalled = 0
+	m.mockPing = func(ctx context.Context) (proto.PingResponse, error) {
+		mockPingCalled += 1
+		return proto.PingResponse{ServerVersion: "9.0.1"}, nil
+	}
+	_, err = listResources(m, mockHTTPReq, types.KindNode)
+	require.True(t, trace.IsNotImplemented(err))
+	require.Equal(t, 1, mockPingCalled)
+
+	// Test v9.1 ping.
+	mockPingCalled = 0
+	m.mockPing = func(ctx context.Context) (proto.PingResponse, error) {
+		mockPingCalled += 1
+		return proto.PingResponse{ServerVersion: "9.1.4"}, nil
+	}
+	m.mockListResources = func(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error) {
+		return nil, nil
+	}
+	_, err = listResources(m, mockHTTPReq, types.KindNode)
+	require.NoError(t, err)
+	require.Equal(t, 1, mockPingCalled)
+}
+
+func TestHandleClusterNodesGetFallback(t *testing.T) {
+	m := &mockedResourceAPIGetter{}
+
+	// Create mock servers.
+	server1, err := types.NewServer("server1", types.KindNode, types.ServerSpecV2{})
+	require.NoError(t, err)
+	server2, err := types.NewServer("server2", types.KindNode, types.ServerSpecV2{})
+	require.NoError(t, err)
+
+	m.mockListResources = func(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error) {
+		return nil, trace.NotImplemented("not implemented")
+	}
+	m.mockGetNodes = func(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]types.Server, error) {
+		return []types.Server{server1, server2}, nil
+	}
+	m.mockPing = func(ctx context.Context) (proto.PingResponse, error) {
+		return proto.PingResponse{ServerVersion: "9.1"}, nil
+	}
+
+	mockHTTPReq, err := http.NewRequest("", "", nil)
+	require.NoError(t, err)
+
+	teleUser, err := types.NewUser("foo")
+	require.NoError(t, err)
+	mockUserRoles := []types.Role{defaultRoleForNewUser(teleUser, "llama")}
+
+	res, err := handleClusterNodesGet(m, mockHTTPReq, "cluster-name", mockUserRoles)
```

Comment:
I would rather test the endpoints than the private methods.

I'd be tempted to bundle these together into one table, if that makes sense, and run tests for versions<9.1 and version>=9.1 as well.

Reply:
i did initially tried to looking into doing it this way, like I knew I needed a similar setup with `web/apiserver_test.go`, but this setup was already so complex and I wasn't sure how to partially mock things out with the auth server, and thought it was faster to just test this way.

Reply:
Fair enough. It looks doable from afar, but I'm sure it gets harder to pull off up close.